gadgets/cmr/CmrRealTimeLAXCineAIAnalysisGadget.cpp

```diff
@@ -77,6 +80,33 @@ namespace Gadgetron {
 
         GDEBUG_STREAM("meas_max_idx_.slice is " << meas_max_idx_.slice);
 
+        // load the model
+        if (this->perform_AI.value())
+        {
+            std::string model_file_name = this->gt_model_home_ + "/" + this->model_file.value();
+            boost::filesystem::path p(model_file_name);
+
+            if (boost::filesystem::exists(p) && boost::filesystem::is_regular_file(p))
+            {
+                GDEBUG_STREAM("model file already exists :  " << model_file_name);
```

Comment:
Could there be concurrency issues here? I'm thinking two concurrent reconstructions try to download the model at the same time. Also it seems like the validation done in get_model is skipped here, but a potentially corrupted file is still left on disk.

Reply:
@johnstairs This is a good point. To get your better opinion, should I used e.g. fcntl.flock to lock the file before writing ? 
A typical usecase is a docker image runs with a gadgetron inside and two connections come in for this same chain.

Reply:
I added a file lock to prevent racing condition to retrieve model file.

Reply:
One problem in the updated version is that you hold the lock while downloading but not while loading. So if two processes (started by separate clients) run different versions of the same chain (they could be passed in from the outside). The first process would download and validate....then process two would overwrite the file because it has a different hash and then process one loads the model, which is now incorrect. I think you may have to bake the hash into the filename or store it in a folder based on the hash of the file. So something like:

```
models --|
             <hash of model file> ---|
                                                  name of model
```
So you download it to a temp location and copy it into that location based on the hash. Then a subsequent process can quickly check if the model it is looking for is there. And you don't start accidentally writing different models to the same name.

Or some variation on that theme....have not totally thought it through, but the way you are doing it now is a step on the way, but I think you could still pretty easily get collisions.